### PR TITLE
add phpstan on level 1

### PR DESCRIPTION
**Pull request by dmaicher**
What do you think about adding phpstan (for now with level=1) to the build? It can catch some simple issues.

Found issues on master were:

```
 ------ ---------------------------------------------------------------------------------------------------------------------------------------------------------------------------- 
  Line   tests/Component/RequestBuilder/ReRankQueryTest.php                                                                                                                          
 ------ ---------------------------------------------------------------------------------------------------------------------------------------------------------------------------- 
         Class Solarium\Tests\QueryType\Select\RequestBuilder\Component\ReRankQueryTest was not found while trying to analyse it - autoloading is probably not configured properly.  
 ------ ---------------------------------------------------------------------------------------------------------------------------------------------------------------------------- 

 ------ ------------------------------------------------------------------------------------------------------------------------------------------------------------------------ 
  Line   tests/Component/RequestBuilder/SpatialTest.php                                                                                                                          
 ------ ------------------------------------------------------------------------------------------------------------------------------------------------------------------------ 
         Class Solarium\Tests\QueryType\Select\RequestBuilder\Component\SpatialTest was not found while trying to analyse it - autoloading is probably not configured properly.  
 ------ ------------------------------------------------------------------------------------------------------------------------------------------------------------------------ 

 ------ ------------------------------------------------------------------------------------------------- 
  Line   tests/Component/Result/MoreLikeThis/ResultTest.php                                               
 ------ ------------------------------------------------------------------------------------------------- 
  18     Access to an undefined property Solarium\Tests\Component\Result\MoreLikeThis\ResultTest::$docs.  
  38     Access to an undefined property Solarium\Tests\Component\Result\MoreLikeThis\ResultTest::$docs.  
  48     Access to an undefined property Solarium\Tests\Component\Result\MoreLikeThis\ResultTest::$docs.  
  53     Access to an undefined property Solarium\Tests\Component\Result\MoreLikeThis\ResultTest::$docs.  
 ------ ------------------------------------------------------------------------------------------------- 

 ------ ---------------------------------------------------------------------------------------------- 
  Line   tests/Component/Result/Suggester/ResultTest.php                                               
 ------ ---------------------------------------------------------------------------------------------- 
  19     Access to an undefined property Solarium\Tests\Component\Result\Suggester\ResultTest::$docs.  
  40     Access to an undefined property Solarium\Tests\Component\Result\Suggester\ResultTest::$docs.  
  50     Access to an undefined property Solarium\Tests\Component\Result\Suggester\ResultTest::$docs.  
  55     Access to an undefined property Solarium\Tests\Component\Result\Suggester\ResultTest::$docs.  
 ------ ---------------------------------------------------------------------------------------------- 

 ------ ---------------------------------------------------------------------------------------- 
  Line   tests/Core/Client/ClientTest.php                                                        
 ------ ---------------------------------------------------------------------------------------- 
  471    Method PHPUnit\Framework\TestCase::createMock() invoked with 2 parameters, 1 required.  
 ------ ---------------------------------------------------------------------------------------- 

 ------ --------------------------------------------------------------------------------------------- 
  Line   tests/Plugin/Loadbalancer/LoadbalancerTest.php                                               
 ------ --------------------------------------------------------------------------------------------- 
  438    Method PHPUnit\Framework\TestCase::expectException() invoked with 2 parameters, 1 required.  
 ------ --------------------------------------------------------------------------------------------- 

 ------ -------------------------------------------------------------------------------------------------------------------------------------------------- 
  Line   tests/QueryType/Graph/QueryTest.php                                                                                                               
 ------ -------------------------------------------------------------------------------------------------------------------------------------------------- 
         Class Solarium\Tests\QueryType\Graph\Query\QueryTest was not found while trying to analyse it - autoloading is probably not configured properly.  
 ------ -------------------------------------------------------------------------------------------------------------------------------------------------- 

 ------ ----------------------------------------------------------------------------------------------------------------------------------------------------------------- 
  Line   tests/QueryType/ManagedResources/QueryTest.php                                                                                                                   
 ------ ----------------------------------------------------------------------------------------------------------------------------------------------------------------- 
         Class Solarium\Tests\QueryType\ManagedResources\Resources\QueryTest was not found while trying to analyse it - autoloading is probably not configured properly.  
 ------ ----------------------------------------------------------------------------------------------------------------------------------------------------------------- 

 ------ ------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------ 
  Line   tests/QueryType/ManagedResources/RequestBuilder/ResourcesTest.php                                                                                                                   
 ------ ------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------ 
         Class Solarium\Tests\QueryType\ManagedResources\Resources\RequestBuilder\ResourcesTest was not found while trying to analyse it - autoloading is probably not configured properly.  
 ------ ------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------ 

 ------ ------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------ 
  Line   tests/QueryType/ManagedResources/RequestBuilder/StopwordsTest.php                                                                                                                   
 ------ ------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------ 
         Class Solarium\Tests\QueryType\ManagedResources\Resources\RequestBuilder\StopwordsTest was not found while trying to analyse it - autoloading is probably not configured properly.  
 ------ ------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------ 

 ------ ----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- 
  Line   tests/QueryType/ManagedResources/RequestBuilder/SynonymsTest.php                                                                                                                   
 ------ ----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- 
         Class Solarium\Tests\QueryType\ManagedResources\Resources\RequestBuilder\SynonymsTest was not found while trying to analyse it - autoloading is probably not configured properly.  
 ------ ----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- 

 ------ --------------------------------------------------------------------------------------- 
  Line   tests/QueryType/Server/Api/RequestBuilderTest.php                                      
 ------ --------------------------------------------------------------------------------------- 
  70     Method PHPUnit\Framework\Assert::arrayHasKey() invoked with 2 parameters, 1 required.  
  85     Method PHPUnit\Framework\Assert::arrayHasKey() invoked with 2 parameters, 1 required.  
  86     Method PHPUnit\Framework\Assert::arrayHasKey() invoked with 2 parameters, 1 required.  
  101    Method PHPUnit\Framework\Assert::arrayHasKey() invoked with 2 parameters, 1 required.  
  102    Method PHPUnit\Framework\Assert::arrayHasKey() invoked with 2 parameters, 1 required.  
  118    Method PHPUnit\Framework\Assert::arrayHasKey() invoked with 2 parameters, 1 required.  
  119    Method PHPUnit\Framework\Assert::arrayHasKey() invoked with 2 parameters, 1 required.  
 ------ --------------------------------------------------------------------------------------- 

 ------ -------------------------------------------------------------------------------------------------------------------------------------------------------- 
  Line   tests/QueryType/Stream/ExpressionTest.php                                                                                                               
 ------ -------------------------------------------------------------------------------------------------------------------------------------------------------- 
         Class Solarium\Tests\QueryType\Stream\Query\ExpressionTest was not found while trying to analyse it - autoloading is probably not configured properly.  
 ------ -------------------------------------------------------------------------------------------------------------------------------------------------------- 

 ------ --------------------------------------------------------------------------------------------------------------------------------------------------- 
  Line   tests/QueryType/Stream/QueryTest.php                                                                                                               
 ------ --------------------------------------------------------------------------------------------------------------------------------------------------- 
         Class Solarium\Tests\QueryType\Stream\Query\QueryTest was not found while trying to analyse it - autoloading is probably not configured properly.  
 ------ --------------------------------------------------------------------------------------------------------------------------------------------------- 

                                                                                                                        
 [ERROR] Found 26 errors                                                                                                
                                                                                                                        
```

### CODE REVIEW

**Comment by thomascorthals** [comment]
An extra pair of (automated) eyes never hurts.

**Comment by wickedOne** [comment]
i'm in favor, but would suggest adding [phpunit](https://github.com/phpstan/phpstan-phpunit) rules while we're at it.

adding [deprecation-rules](https://github.com/phpstan/phpstan-deprecation-rules) might be usefull as well as there are some dependencies with likely deprecations, but i'm not sure you're able to configure it to ignore deprecation warnings thrown from within your own library (which we have to when we want to deprecate features ourselves)

File: composer.json [hunk 1]
```diff
@@ -24,6 +24,7 @@
         "nyholm/psr7": "^1.2",
         "php-coveralls/php-coveralls": "^2.1",
         "php-http/guzzle6-adapter": "^2.0",
+        "phpstan/phpstan": "^0.12.18",
```
Comment (wickedOne) [inline composer.json:27]: might want to change this to ``"^0.12"``

Comment (dmaicher) [reply composer.json:27]: Yeah composer chose this constraint with `composer req phpstan/phpstan --dev` :yum:

Comment (wickedOne) [reply composer.json:27]: yes composer behaves a bit different when packages have a version < 1
changing it to "^0.12" will at least ensure we're up to date with the patches

**Comment by dmaicher** [comment]
> i'm in favor, but would suggest adding [phpunit](https://github.com/phpstan/phpstan-phpunit) rules while we're at it.

I added it :wink: 84 errors fixed :see_no_evil:  Luckily more than half could be fixed by regex search&replace :smile: 
 
> adding [deprecation-rules](https://github.com/phpstan/phpstan-deprecation-rules) might be usefull as well as there are some dependencies with likely deprecations, but i'm not sure you're able to configure it to ignore deprecation warnings thrown from within your own library (which we have to when we want to deprecate features ourselves)

Never used that one. Could also be investigated once this one is merged I guess?

File: composer.json (hunk 1 above)
Comment (dmaicher) [reply composer.json]: Changed it.

Although: Isn't `^0.12.18` the same as `>=0.12.18, <0.13`?

Comment (wickedOne) [reply composer.json]: yes you're right :|
i was confused and thought it would stick to the patches, but sticks to the minor for pre 1.0 releases...

**Comment by wickedOne** [comment]
> Never used that one. Could also be investigated once this one is merged I guess?

sure, new rules can always be added.

**Pull request merged**